src/providers/MeetingProvider/MeetingManager.ts

```diff
@@ -214,29 +166,77 @@ export class MeetingManager implements AudioVideoObserver {
     this.audioVideoObservers = {};
   }
 
-  async join({
-    meetingInfo,
-    attendeeInfo,
-    deviceLabels = DeviceLabels.AudioAndVideo,
-    eventReporter,
-    meetingManagerConfig,
-  }: MeetingJoinData) {
-    if (meetingManagerConfig) {
-      this.meetingManagerConfig = meetingManagerConfig;
-    }
-    this.configuration = new MeetingSessionConfiguration(
-      meetingInfo,
-      attendeeInfo
+  async join(meetingJoinData: MeetingJoinData): Promise<void> {
+    const { meetingSessionConfiguration } = meetingJoinData;
+    const {
+      deviceLabels,
+      eventReporter,
+      logger,
+      enableWebAudio,
+      activeSpeakerPolicy,
+    } = this.setupMeetingManager(meetingJoinData, meetingSessionConfiguration);
+    this.meetingSessionConfiguration = meetingSessionConfiguration;
+    this.meetingId = this.meetingSessionConfiguration.meetingId;
+    this.logger = logger;
+
+    const deviceController = new DefaultDeviceController(logger, {
+      enableWebAudio: enableWebAudio,
+    });
+    this.meetingSession = new DefaultMeetingSession(
+      meetingSessionConfiguration,
+      logger,
+      deviceController,
+      eventReporter
     );
 
-    this.meetingRegion = meetingInfo.MediaRegion;
-    this.meetingId = this.configuration.meetingId;
+    this.audioVideo = this.meetingSession.audioVideo;
 
-    if (eventReporter) {
-      this.eventReporter = eventReporter;
-    }
+    // When an attendee leaves, we remove AudioVideoObservers and nullify AudioVideoFacade and MeetingSession object.
+    // This results into missing few meeting events triggered with audioVideoDidStop such as meetingEnded, meetingFailed
+    // and meetingStartFailed. We may also loose audioInputUnselected and videoInputUnselected events as we choose null
+    // devices when an attendee leaves. When a new AudioVideoFacade object is created remove and re-add the
+    // eventDidReceive observer which wont leak.
+    this.audioVideo.removeObserver(this.eventDidReceiveRef);
+    this.audioVideo.addObserver(this.eventDidReceiveRef);
 
-    await this.initializeMeetingSession(this.configuration, deviceLabels);
+    this.setupAudioVideoObservers();
+    this.setupDeviceLabelTrigger(deviceLabels);
+    await this.listAndSelectDevices(deviceLabels);
+    this.publishAudioVideo();
+    this.setupActiveSpeakerDetection(activeSpeakerPolicy);
+    this.meetingStatus = MeetingStatus.Loading;
+    this.publishMeetingStatus();
+  }
+
+  private setupMeetingManager(
```

Comment:
This method does not setup `meetingManager`, it actually just parses the `meetingJoinData`.
And if the naming is `setupMeetingManager` which mostly means it returns void. But actually it returns the parsed `meetingJoinData`.

So these indicate the naming of this method is not right. Need to rename this method.

Reply:
I was not really that confident of this name. I have renamed it to `parseMeetingManagerJoinData`. Let me know if that works. I think that is more akin to what the method does.

Reply:
Renamed to `parseJoinParams`.